README
This is pmtools -- a suite of small programs to help manage modules.
The names are totally preliminary, and in fact, so is the code.  We follow
the "keep it small" notion of many tiny tools each doing one thing well,
eschewing giant megatools with millions of options.

Tom Christiansen
tchrist@perl.com


=========================================================

pmpath - show the module's full path

    $ pmpath Carp
    /usr/local/devperl/lib/5.00554/Carp.pm

=========================================================

pmvers - get a module version number

    $ pmvers CGI
    2.46

    $ perl5.00404 -S pmvers CGI
    2.42

    $ devperl -S pmvers CGI
    2.46

=========================================================

pmdesc  - get a module description

    $ pmdesc IO::Dir
    IO::Dir (1.03) - supply object methods for directory handles

=========================================================

pmall  - get all installed modules pmdesc descriptions 

    $ pmall
    Pod::Html (1.01) - module to convert pod files to HTML
    Config - access Perl configuration information
    DB_File (1.61) - Perl5 access to Berkeley DB
    DynaLoader (1.03) - Dynamically load C libraries into Perl code
    Fcntl (1.03) - load the C Fcntl.h defines
    File::Basename (2.6) - split a pathname into pieces
    (etc)

  This one is a bit crufty.  It takes some options.
  Check out the source.

=========================================================

pmcheck   - check that Perl is set up correctly for Perl modules

    $ pmcheck
    '/usr/lib/perl5/site_perl/5.8' is in @INC but does not exist

=========================================================

pmdirs    - print the perl module path, newline separated

    $ pmdirs
    /home/tchrist/perllib/i686-linux
    /home/tchrist/perllib
    /usr/local/devperl/lib/5.00554/i686-linux
    /usr/local/devperl/lib/5.00554
    /usr/local/devperl/lib/site_perl/5.00554/i686-linux
    /usr/local/devperl/lib/site_perl/5.00554
    .

    $ filsperl -S pmdirs
    /home/tchrist/perllib
    /usr/local/filsperl/lib/5.00554/i686-linux-thread
    /usr/local/filsperl/lib/5.00554
    /usr/local/filsperl/lib/site_perl/5.00554/i686-linux-thread
    /usr/local/filsperl/lib/site_perl/5.00554
    .

=========================================================

plxload - show what files a given program loads at compile time

    $ plxload perldoc
    /usr/local/devperl/lib/5.00554/Exporter.pm
    /usr/local/devperl/lib/5.00554/strict.pm
    /usr/local/devperl/lib/5.00554/vars.pm
    /usr/local/devperl/lib/5.00554/i686-linux/Config.pm
    /usr/local/devperl/lib/5.00554/Getopt/Std.pm

    $ plxload /usr/src/perl5.005_54/installhtml
    /usr/local/devperl/lib/5.00554/Carp.pm
    /usr/local/devperl/lib/5.00554/Exporter.pm
    /usr/local/devperl/lib/5.00554/auto/Getopt/Long/autosplit.ix
    /usr/local/devperl/lib/5.00554/strict.pm
    /usr/local/devperl/lib/5.00554/vars.pm
    /usr/local/devperl/lib/5.00554/Pod/Functions.pm
    /usr/local/devperl/lib/5.00554/Getopt/Long.pm
    /usr/local/devperl/lib/5.00554/i686-linux/Config.pm
    /usr/local/devperl/lib/5.00554/lib.pm
    /home/tchrist/perllib/Pod/Html.pm
    /usr/local/devperl/lib/5.00554/Cwd.pm
    /usr/local/devperl/lib/5.00554/AutoLoader.pm

=========================================================

pmload - show what files a given module loads at compile time

    $ pmload IO::Handle
    /usr/local/devperl/lib/5.00554/Exporter.pm
    /usr/local/devperl/lib/5.00554/Carp.pm
    /usr/local/devperl/lib/5.00554/strict.pm
    /usr/local/devperl/lib/5.00554/vars.pm
    /usr/local/devperl/lib/5.00554/i686-linux/DynaLoader.pm
    /usr/local/devperl/lib/5.00554/i686-linux/IO/Handle.pm
    /usr/local/devperl/lib/5.00554/Symbol.pm
    /usr/local/devperl/lib/5.00554/i686-linux/IO/File.pm
    /usr/local/devperl/lib/5.00554/SelectSaver.pm
    /usr/local/devperl/lib/5.00554/i686-linux/Fcntl.pm
    /usr/local/devperl/lib/5.00554/AutoLoader.pm
    /usr/local/devperl/lib/5.00554/i686-linux/IO.pm
    /usr/local/devperl/lib/5.00554/i686-linux/IO/Seekable.pm

    $ cat `pmload IO::File` | wc -l
       3131

    $ pmload Tk
    /usr/lib/perl5/site_perl/Tk/Pretty.pm
    /usr/lib/perl5/Symbol.pm
    /usr/lib/perl5/site_perl/Tk/Frame.pm
    /usr/lib/perl5/site_perl/Tk/Toplevel.pm
    /usr/lib/perl5/strict.pm
    /usr/lib/perl5/Exporter.pm
    /usr/lib/perl5/vars.pm
    /usr/lib/perl5/site_perl/auto/Tk/Wm/autosplit.ix
    /usr/lib/perl5/site_perl/auto/Tk/Widget/autosplit.ix
    /usr/lib/perl5/site_perl/Tk.pm
    /usr/lib/perl5/i386-linux/5.00404/DynaLoader.pm
    /usr/lib/perl5/site_perl/auto/Tk/Frame/autosplit.ix
    /usr/lib/perl5/site_perl/auto/Tk/Toplevel/autosplit.ix
    /usr/lib/perl5/Carp.pm
    /usr/lib/perl5/site_perl/auto/Tk/autosplit.ix
    /usr/lib/perl5/site_perl/Tk/CmdLine.pm
    /usr/lib/perl5/site_perl/Tk/MainWindow.pm
    /usr/lib/perl5/site_perl/Tk/Submethods.pm
    /usr/lib/perl5/site_perl/Tk/Configure.pm
    /usr/lib/perl5/AutoLoader.pm
    /usr/lib/perl5/site_perl/Tk/Derived.pm
    /usr/lib/perl5/site_perl/Tk/Image.pm
    /usr/lib/perl5/site_perl/Tk/Wm.pm
    /usr/lib/perl5/site_perl/Tk/Widget.pm

=========================================================

pmexp - show a module's exports

    $ pmexp Text::ParseWords
    Text::ParseWords automatically exports shellwords, quotewords, nested_quotewords, and parse_line
    Text::ParseWords optionally exports old_shellwords

    $ pmexp Text::Wrap
    Text::Wrap automatically exports wrap and fill
    Text::Wrap optionally exports $columns, $break, and $huge

    $ pmexp Fcntl
    Fcntl automatically exports FD_CLOEXEC, F_DUPFD, F_EXLCK, F_GETFD, F_GETFL, F_GETLK, F_GETLK64, F_GETOWN, F_POSIX, F_RDLCK, F_SETFD, F_SETFL, F_SETLK, F_SETLK64, F_SETLKW, F_SETLKW64, F_SETOWN, F_SHLCK, F_UNLCK, F_WRLCK, O_ACCMODE, O_APPEND, O_ASYNC, O_BINARY, O_CREAT, O_DEFER, O_DSYNC, O_EXCL, O_EXLOCK, O_LARGEFILE, O_NDELAY, O_NOCTTY, O_NONBLOCK, O_RDONLY, O_RDWR, O_RSYNC, O_SHLOCK, O_SYNC, O_TEXT, O_TRUNC, and O_WRONLY
    Fcntl optionally exports FAPPEND, FASYNC, FCREAT, FDEFER, FEXCL, FNDELAY, FNONBLOCK, FSYNC, FTRUNC, LOCK_EX, LOCK_NB, LOCK_SH, and LOCK_UN
    Fcntl export tag `Fcompat' includes FAPPEND, FASYNC, FCREAT, FDEFER, FEXCL, FNDELAY, FNONBLOCK, FSYNC, and FTRUNC
    Fcntl export tag `flock' includes LOCK_SH, LOCK_EX, LOCK_NB, and LOCK_UN

=========================================================

pminst  - find what's installed

    $ pminst
    (lists all installed modules)

    $ pminst Carp
    CGI::Carp
    Carp

    $ pminst ^IO::
    IO::Socket::INET
    IO::Socket::UNIX
    IO::Select
    IO::Socket
    IO::Poll
    IO::Handle
    IO::Pipe
    IO::Seekable
    IO::Dir
    IO::File

    $ pminst '(?i)io'
    IO::Socket::INET
    IO::Socket::UNIX
    IO::Select
    IO::Socket
    IO::Poll
    IO::Handle
    IO::Pipe
    IO::Seekable
    IO::Dir
    IO::File
    IO
    Pod::Functions

  The -s flag provides output with the directory separated
  by a space:

    $ pminst -s | sort +1
    (lists all modules, sorted by name, but with where they 
     came from)

    $ perl5.00404 -S pminst -s IO
    /usr/lib/perl5/i386-linux/5.00404 IO::File
    /usr/lib/perl5/i386-linux/5.00404 IO::Handle
    /usr/lib/perl5/i386-linux/5.00404 IO::Pipe
    /usr/lib/perl5/i386-linux/5.00404 IO::Seekable
    /usr/lib/perl5/i386-linux/5.00404 IO::Select
    /usr/lib/perl5/i386-linux/5.00404 IO::Socket
    /usr/lib/perl5/i386-linux/5.00404 IO
    /usr/lib/perl5/site_perl LWP::IO
    /usr/lib/perl5/site_perl LWP::TkIO
    /usr/lib/perl5/site_perl Tk::HTML::IO
    /usr/lib/perl5/site_perl Tk::IO
    /usr/lib/perl5/site_perl IO::Stringy
    /usr/lib/perl5/site_perl IO::Wrap
    /usr/lib/perl5/site_perl IO::ScalarArray
    /usr/lib/perl5/site_perl IO::Scalar
    /usr/lib/perl5/site_perl IO::Lines
    /usr/lib/perl5/site_perl IO::WrapTie
    /usr/lib/perl5/site_perl IO::AtomicFile

  The -l flag gives full paths:

    $ filsperl -S pminst -l Thread
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread/Queue.pm
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread/Semaphore.pm
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread/Signal.pm
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread/Specific.pm
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread.pm


=========================================================

pmeth - list a class's methods, recursively

    $ pmeth IO::Socket
    AF_INET
    AF_UNIX
    INADDR_ANY
    INADDR_BROADCAST
    INADDR_LOOPBACK
    INADDR_NONE
    SOCK_DGRAM
    SOCK_RAW
    SOCK_STREAM
    accept
    bind
    carp
    confess
    configure
    connect
    connected
    croak
    getsockopt
    import
    inet_aton
    inet_ntoa
    listen
    new
    pack_sockaddr_in
    pack_sockaddr_un
    peername
    protocol
    recv
    register_domain
    send
    setsockopt
    shutdown
    sockaddr_in
    sockaddr_un
    sockdomain
    socket
    socketpair
    sockname
    sockopt
    socktype
    timeout
    unpack_sockaddr_in
    unpack_sockaddr_un
    DESTROY via IO::Handle
    SEEK_CUR via IO::Handle
    SEEK_END via IO::Handle
    SEEK_SET via IO::Handle
    _IOFBF via IO::Handle
    _IOLBF via IO::Handle
    _IONBF via IO::Handle
    _open_mode_string via IO::Handle
    autoflush via IO::Handle
    blocking via IO::Handle
    [overridden] carp via IO::Handle
    clearerr via IO::Handle
    close via IO::Handle
    [overridden] confess via IO::Handle
    constant via IO::Handle
    [overridden] croak via IO::Handle
    eof via IO::Handle
    error via IO::Handle
    fcntl via IO::Handle
    fdopen via IO::Handle
    fileno via IO::Handle
    flush via IO::Handle
    format_formfeed via IO::Handle
    format_line_break_characters via IO::Handle
    format_lines_left via IO::Handle
    format_lines_per_page via IO::Handle
    format_name via IO::Handle
    format_page_number via IO::Handle
    format_top_name via IO::Handle
    format_write via IO::Handle
    formline via IO::Handle
    gensym via IO::Handle
    getc via IO::Handle
    getline via IO::Handle
    getlines via IO::Handle
    gets via IO::Handle
    input_line_number via IO::Handle
    input_record_separator via IO::Handle
    ioctl via IO::Handle
    [overridden] new via IO::Handle
    new_from_fd via IO::Handle
    opened via IO::Handle
    output_field_separator via IO::Handle
    output_record_separator via IO::Handle
    print via IO::Handle
    printf via IO::Handle
    printflush via IO::Handle
    qualify via IO::Handle
    qualify_to_ref via IO::Handle
    read via IO::Handle
    setbuf via IO::Handle
    setvbuf via IO::Handle
    stat via IO::Handle
    sync via IO::Handle
    sysread via IO::Handle
    syswrite via IO::Handle
    truncate via IO::Handle
    ungensym via IO::Handle
    ungetc via IO::Handle
    untaint via IO::Handle
    write via IO::Handle
    _push_tags via Exporter via IO::Handle
    export via Exporter via IO::Handle
    export_fail via Exporter via IO::Handle
    export_ok_tags via Exporter via IO::Handle
    export_tags via Exporter via IO::Handle
    export_to_level via Exporter via IO::Handle
    [overridden] import via Exporter via IO::Handle
    require_version via Exporter via IO::Handle
    VERSION via UNIVERSAL
    can via UNIVERSAL
    [overridden] import via UNIVERSAL
    isa via UNIVERSAL

You might want to send that through grep or grep -v looking
for "via" or "overridden".

=========================================================

pmls  - long list the module path

    $ pmls CGI
    -r--r--r--   1 root     root       190901 Dec  6 03:19
		/usr/local/devperl/lib/5.00554/CGI.pm

This is mostly here for people too lazy to type
    $ ls -l `pmpath CGI` 

=========================================================

pmcat  - cat the module source through your pager

    $ pmcat CGI

This is mostly here for people too lazy to type
    $ more `pmpath CGI` 

=========================================================

pman - show the module's pod docs

    $ pman CGI
    $ pman Curses

This is mostly here for people too lazy to type
    $ pod2text `pmpath CGI` | more

=========================================================

pmfunc - show a function source code from a module

    $ pmfunc Cwd::getcwd
    sub getcwd
    {
	abs_path('.');
    }

This is mostly here for people who are too lazy to type

    sed '/^sub getcwd/,/}/p' `pmpath Cwd`
or
    perl -ne 'print if /^sub\s+getcwd\b/ .. /}/' `pmpath Cwd`

=========================================================

podgrep - grep in pods of a file

  Flags:
      -i means case insensitive match
      -p means page output 
      -f means format output 
      -h means check for matches in pod =head and =item headers alone,
         and to keep printing podagraphs until the next header is found.

    $ podgrep mail `pmpath CGI`
    (prints out podagraphs from the CGI.pm manpage that mention mail)

    $ podgrep -i destructor `sitepods`
    (prints out podagraphs that mention destructors in the 
     site-installed pods)

    $ podgrep -i 'type.?glob' `stdpods`

    (prints out podagraphs that mention typeglob in the
     standard pods)

    $ podgrep -hpfi "lock" `faqpods`

    (prints out all podagraphs with "lock" in the headers
    case-insensitively, then then formats these with pod2text, then
    shows them in the pager with matches high-lighted)

    $ podgrep -fh seek `podpath perlfunc`
    (prints out and formats podagraphs from the standard perlfunc manpage
    whose headers or items contain "seek".)

=========================================================

pfcat - show pods from perlfunc

(uses podgrep, uses those options)

    $ pfcat seek

is like 

    $ podgrep -h seek `podpath perlfunc`

Whereas

    $ pfcat -pf seek

is like

    $ podgrep -pfh seek `podpath perlfunc`

which is like

    $ podgrep -h seek `podpath perlfunc` | podtext | $PAGER +/seek


=========================================================

podtoc - list table of contents of a podpage

    $ podtoc `pmpath CGI`
     NAME
     SYNOPSIS
     ABSTRACT
     DESCRIPTION
	 PROGRAMMING STYLE
	 CALLING CGI.PM ROUTINES
	 *  1. Use another name for the argument, if one is available.  For
	    example, -value is an alias for -values.
	 *  2. Change the capitalization, e.g. -Values
    (etc)

=========================================================

podpath - show full path of pod file

(like pmpath, but works on standard manpages, too)

    $ podpath Cwd
    /usr/local/devperl/lib/5.00554/Cwd.pm

    $ devperl -S podpath perlfunc
    /usr/local/devperl/lib/5.00554/pod/perlfunc.pod

    $ oldperl -S podpath IO::Handle
    /usr/lib/perl5/i386-linux/5.00404/IO/Handle.pm

    $ filsperl -S podpath Thread
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/Thread.pm


=========================================================

pods - list all standard pods and module pods

    $ pods

=========================================================

sitepods - list only pods in site_perl directories

    $ sitepods
    /usr/local/devperl/lib/site_perl/5.00554/i686-linux/XML/Parser/Expat.pm
    /usr/local/devperl/lib/site_perl/5.00554/i686-linux/XML/Parser.pm


=========================================================

basepods - list only normal "man-page" style pods

    $ basepods | grep delt
    /usr/local/devperl/lib/5.00554/pod/perl5004delta.pod
    /usr/local/devperl/lib/5.00554/pod/perl5005delta.pod
    /usr/local/devperl/lib/5.00554/pod/perldelta.pod


=========================================================

faqpods - list only faq pods

=========================================================

modpods - all module pods, including site_perl ones

=========================================================

stdpods - list standard pods, not site_perl ones


Changes
v2.0.0  2014-02-03 17:59:29 EST  Mark Leighton Fisher

  - Bumped version to v2.00, because of the major architectural
    change -- the 'pmtools' module now contains the routine
    new_pod_iterator(), which creates an iterator for all of
    the variants of a POD's name ('LIB/MODULE.pm',
    'LIB/MODULE.pod', and 'LIB/pod/MODULE.pod'). This change
    supports the use case of when the POD is separate from its
    corresponding Perl module file (inspired by an upcoming
    Debian release).
  - Fixed pman and podpath to use new_pod_iterator().
  - Switched to the more-standard 3-number versioning scheme.

v1.54  2013-10-19 14:21:26 EST  Mark Leighton Fisher

  - Updated $VERSION in the modules and documented it too :)!

v1.53  2013-10-18 17:31:35 EST  Mark Leighton Fisher

  - Tried to update $VERSION in the modules.

v1.52  2013-09-28 19:00:47 EST  Mark Leighton Fisher

  - Updated $VERSION in the modules.

v1.51  2013-09-27 06:19:01 EST  Mark Leighton Fisher

  - NOTE: These are all fixes in the tests.
  - Fixed rt.cpan.org 88843, "t/pmall.t takes a long time" by using
    a head(1) clone (t/head.pl) to grab just the first 100 lines
    of pmall output, then matching against the general form of
    pmall output rather than specifically looking for Scalar::Util.
  - Fixed the tests to use Tie::Hash (pure Perl) instead of
    Scalar::Util, as there may or may not be a ...Scalar/Util.pm
    depending on how Perl was packaged and installed (Scalar::Util
    can exist as a pure Perl module and (more powerfully) as a mix
    of Perl and XS).
  - Fixed t/pmfunc.t and t/pmcheck.t to just check $? != 127 (i.e.
    the program ran in some way) rather than looking for a specific
    exit code. pmfunc does not work again yet, and pmcheck may or
    may not return an error depending whether your Perl configuration
    is correct.
  - Fixed t/pman.t, t/pmcat.t, and t/pmdesc.t to use the Carp module
    because Carp always has POD (Tie::Hash may not depending on
    the Perl version, and see above for Scalar::Util problems).
    
v1.50  2013-09-20 17:00:33 EST  Mark Leighton Fisher

  - This is the "Testing" release of pmtools! All of the
    pmtools now have tests.
  - Fixed so all programs and modules use a fixed value
    (literal) for $VERSION. This seems to work better with
    the CPAN toolchain. As I also updated all of the
    version numbers, this will fix rt.cpan.org
    "#88534: pmtools 1.40: Still has "1.30" in pmtools.pm"
  - Fixed so all tests add the pmtools to the PATH, whether
    or not the test actually needs it. This should fix
    rt.cpan.org "#88780: test failure".
  - Fixed pfcat documentation to always refer to "pfcat".
  - Fixed rt.cpan.org "#85352: typo fixes".

v1.40  2013-09-04 16:53:36 EST  Mark Leighton Fisher

  - This is the "Kwalitee" release of pmtools! All of the
    Test::Kwalitee and Test::Kwalitee::Extra tests now pass.
  - I have (briefly) tested this release on Debian 7.1
    with Perl 5.8.8, 5.14.2, and 5.18.1.
  - Please note that there are currently only tests for
    basepods and faqpods. My current plan is add tests for
    all other programs and modules in the next release.
  - This is the first release where I used Dist::Zilla to
    create the release. A special thanks to David Golden
    for bundling some many useful Dist::Zilla plugins
    together (Dist::Zilla::PluginBundle::DAGOLDEN).
  - All programs and modules now 'use strict' and
    'use warnings' (part of "Kwalitee").
  - Licensing information is now in the LICENSE section
    for programs and modules (part of "Kwalitee').
  - There are now LICENSE, MYMETA.json, META.yml, and
    MYMETA.yml files (part of "Kwalitee").
  - Programs are now in the standard location 'bin'.
  - Devel::Loaded is now in the standard location
    lib/Devel/Loaded.pm.
  - There is now a 'pmtools' module (lib/pmtools.pm).
  - 'pmtools' is now the source of the version number for
    all other programs and modules in pmtools. This should
    satisfy dsteinbrunner's pull request of April 27, 2013.
  - 'pminclude' lists the directories in @INC. This
    replaces the poorly-named 'pmdirs' program.
  - Documentation for Devel::Loaded has been fixed. This
    should satisfy neilbowers' pull request of December 21,
    2012 and neilbowers' pull request of July 16, 2013.
  - I've fixed some permissions and such that should satisfy
    dolmen's pull request of December 25, 2012.
  - INTERNAL DETAIL: This version was developed on a
    separate branch rather than 'master' (some Git
    goodness).

v1.30  2013-01-18 17:24:19 EST  Mark Leighton Fisher

  - Moved to Git for source control.
  - Added the GitHub repo
    https://github.com/markleightonfisher/pmtools.
  - Fixed Devel/Loaded.pm so it is not added to the
    output. Thanks to NEILB@cpan.org for spotting this
    bug (a modern Perls bug).
  - Patched pmeth to flag Perl constants as such (thanks
    to jpierce@cpan.org for the patch). XS constants
    are not yet marked (some C-fu needs to be applied).
  - Patched basepods to use 'pods' as the directory for
    the base PODs (modern Perls use 'pods'). Alas, this
    appears to be inspired by a CPANgnome, as there was
    no Requestor on the ticket. :(

v1.20  2012-12-07 15:25:50 EST  Mark Leighton Fisher

  - Patched pminst so only unique package names are printed.
    Thanks to Matthew Persico for the patch. (Honorable
    mention goes to Andrew Pam, but Matthew's patch is
    smaller and more robust as it handles an @INC with
    directories in any order.)
  - Fixed pminst to ignore non-existent directories.
  - Added pmcheck, a tool to check that Perl is set up
    correctly for use with Perl modules. Currently,
    pmcheck only verifies that everything in @INC is an
    existing, readable directory.

v1.10  2008-01-18 06:57:21 EST  Mark Leighton Fisher

  - Added copyrights in my name as well as Tom Christiansen.
  - Changed to dual-license like Perl: GNU and Perl Artistic.

lib/pmtools.pm
package pmtools;

use strict;
use warnings;

our $VERSION = '2.0';

sub new_pod_iterator {
    my ($dir, $module) = @_;

    my $index = 0;
    my @pod_dirs = (
        "$dir/pod/$module.pod",
        "$dir/$module.pod",
        "$dir/$module.pm",
    );

    return sub {
        if ($index >= scalar @pod_dirs) {
            return undef;
        }

	return $pod_dirs[$index++];
    };
}

1;


# ------ From Tom Christiansen's original README file for pmtools:
# This is pmtools -- a suite of small programs to help manage modules.
# The names are totally preliminary, and in fact, so is the code.  We follow
# the "keep it small" notion of many tiny tools each doing one thing well,
# eschewing giant megatools with millions of options.
#
# Tom Christiansen
# tchrist@perl.com

=head1 NAME

pmtools - Perl Module Tools

=head1 SYNOPSIS

pmdesc module ...
pmpath module ...
pmvers module ...
...
sitepods
...

=head1 DESCRIPTION

Perl Modules Tools (pmtools) is a suite of tools designed
originally by Tom Christiansen.

For each module name given as an argument, produces its full path on
the standard output, one per line.

=head1 EXAMPLES

    $ pmpath CGI
    /usr/local/devperl/lib/5.00554/CGI.pm

    $ filsperl -S pmpath IO::Socket CGI::Carp
    /usr/local/filsperl/lib/5.00554/i686-linux-thread/IO/Socket.pm
    /usr/local/filsperl/lib/5.00554/CGI/Carp.pm

    $ oldperl -S pmpath CGI CGI::Imagemap
    /usr/lib/perl5/CGI.pm
    /usr/lib/perl5/site_perl/CGI/Imagemap.pm

=head1 SEE ALSO

pmdesc(1),
pmvers(1),
pmcat(1).

=head1 AUTHORS and COPYRIGHTS

Copyright (C) 1999 Tom Christiansen.

Copyright (C) 2006-2013 Mark Leighton Fisher.

=head1 LICENSE

This is free software; you can redistribute it and/or modify it
under the terms of either:
(a) the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or
(b) the Perl "Artistic License".
(This is the Perl 5 licensing scheme.)

Please note this is a change from the
original pmtools-1.00 (still available on CPAN),
as pmtools-1.00 were licensed only under the
Perl "Artistic License".


lib/Devel/Loaded.pm
# Loaded.pm -- show what files were loaded 
package Devel::Loaded;

# ABSTRACT: pmtools - Perl Module Tools

# ------ pragmas
use strict;
use warnings;
use File::Spec;

our $VERSION = '2.0';

# ----- define variable
my $path  = undef;	# current module path
our %Seen = ();		# track whether we've seen this module before

BEGIN { %Seen = %INC } 

END { 
    # get last part of Devel::Loaded path; handle File::Spec->canonpath() removing '.' from '.pm'
    my $devel_loaded = File::Spec->catfile('Devel', 'Loaded.pm');
    if ($devel_loaded =~ m/Loadedpm$/) {
        $devel_loaded =~ s/Loadedpm$/Loaded.pm/;
    }

    # delete the matching absolute path from %INC
    # NOTE: this will fail if you also have directories like ".../MyOwnDevel/Loaded.pm"
    my $inc_devel_loaded = "";
    foreach $path (keys(%INC)) {
       if ($path =~ m/$devel_loaded$/) {
            $inc_devel_loaded = $path;
            last;
        } 
    }
    if ($inc_devel_loaded ne "") {
        #delete $INC{$inc_devel_loaded};
    }

    for my $path (values %INC) {
       print "$path\n" unless $Seen{$path};
    }

}

1;

__END__

=head1 NAME

Devel::Loaded - Post-execution dump of loaded modules

=head1 SYNOPSIS

    perl -MDevel::Loaded programname 2>/dev/null

=head1 DESCRIPTION

The Devel::Loaded module installs an at-exit handler to generate a dump of
all the module files used by a Perl program.  If used in conjunction with
a I<perl -c>, you find those files loaded in at compile time with C<use>.
If you are willing to wait until after the program runs, you can get
them all.

=head1 EXAMPLES

This is compile-time only:

    $ perl -MDevel::Loaded perldoc 2>/dev/null
    /usr/local/devperl/lib/5.00554/Exporter.pm
    /usr/local/devperl/lib/5.00554/strict.pm
    /usr/local/devperl/lib/5.00554/vars.pm
    /usr/local/devperl/lib/5.00554/i686-linux/Config.pm
    /usr/local/devperl/lib/5.00554/Getopt/Std.pm

This will also catch run-time loads:

    #!/usr/bin/perl
    use Devel::Loaded;
    ...

=head1 SEE ALSO

The I<plxload> and the I<pmload> programs, which use
this technique.

Neil Bowers has written a L<review|http://neilb.org/reviews/dependencies.html> of
Perl modules for getting dependency information (26 of them at the time of writing). 

=head1 AUTHORS and COPYRIGHTS

Copyright (C) 1999 Tom Christiansen.

Copyright (C) 2006-2013 Mark Leighton Fisher.

=head1 LICENSE

This is free software; you can redistribute it and/or modify it
under the terms of either:
(a) the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or
(b) the Perl "Artistic License".
(This is the Perl 5 licensing scheme.)

Please note this is a change from the
original pmtools-1.00 (still available on CPAN),
as pmtools-1.00 were licensed only under the
Perl "Artistic License".
